brave/utility/tor/tor_impl.cc

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) 2018 The Brave Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "brave/utility/tor/tor_impl.h"
+
+#include <utility>
+
+#include "base/command_line.h"
+#include "base/process/kill.h"
+#include "base/process/launch.h"
+#include "base/single_thread_task_runner.h"
+#include "base/task_scheduler/post_task.h"
+#include "base/threading/thread.h"
+#include "base/threading/thread_task_runner_handle.h"
+
+#if defined(OS_POSIX)
+int pipehack[2];
+
+static void SIGCHLDHandler(int signal) {
+  pid_t p;
+  int status;
+
+  if ((p = waitpid(-1, &status, WNOHANG | WUNTRACED)) != -1) {
+    if (WIFSIGNALED(status) || WCOREDUMP(status)) {
+      write(pipehack[1], &p, sizeof(pid_t));
+    } else if (WIFEXITED(status)) {
+      // TODO(darkdh): Handle "Address already in use" issue. Not notifying as
+      // workaroud for now to prevent infinitely relaunch
```

Comment:
Eventually we should avoid restarting until we have successfully opened a control channel, and also rate-limit the restarts, and send the output of the tor process to the console along with a log message about the status.

Actually, this is a motivation for calling waitpid in the thread rather than the signal handler: easier to report a log message about the status too.

Reply:
fixed in fea7dca